### PR TITLE
Switch to nonmonotone linesearch

### PR DESCRIPTION
**Pull request by ManifoldFR**
Switch to the nonmonotone linesearch

small tweak of ur10_ballistic

acrobot: change default params

HistoryCallback: pass solver pointer as argument to templated ctor

examples: python: change HistoryCallback() calls

re-tune cartpole.py

update quadrotor_obstacles.py

Add separate mu_init for non-bound constrained setting

[utils/plotting] some enhancements to plotting convergence diagrams

[examples] re-tune some problems, plot convergence

Update plotting.py again:

- decide labels automatically
- only add "AL iters" label when there are 2 iters

update solo_jump.py

solo_jump : greatly increase joint-space weight

solo_jump : add spinning

solo_jump : flight phase target irrelevant

solo_jump : spin further, fix landing fn

ur10_ballistic : load_projectile_model() can take free flyer or not

+ adjust rigid constraint model factory, pass contact type

[examples] ur10_ballistic.py : sanely distinguish original model nq, nv from model with projectile's nq, nv

plotting.py : get legend first before adding other labels

ask for tighter tolerance on cartpole

quadrotor_obstacles : use the plotting util

### CODE REVIEW

**Comment by ManifoldFR** [comment]
This change is part of the following stack:

- #242
    - #236
        - #244 ◀

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

**Comment by ManifoldFR** [comment]
This PR is finally ready for review.

File: tests/python/test_solver.py [hunk 1]
```diff
@@ -60,7 +60,9 @@ def test_no_node():
     solver.setup(problem)
 
 
-@pytest.mark.parametrize("strategy", [aligator.SA_FILTER, aligator.SA_LINESEARCH])
+@pytest.mark.parametrize(
```
Comment (edantec) [inline tests/python/test_solver.py:63]: SA_LINESEARCH_ARMIJO is not tested here

**Review by edantec — COMMENTED** [review]
I have a few questions but otherwise the PR seems ok

File: include/aligator/core/linesearch-nonmonotone.hpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  T run(fun_t f, T phi0, T dphi0, T &a_opt);
+  NonmonotoneLinesearch(const Options &options);
+
+  void reset() {
+    mov_avg = T(0.);
+    avg_weight = T(0.);
+  }
+
+  /// Weight for moving average
+  T avg_eta = 0.85;
```
[the remark this answers is not in the document]
Comment (ManifoldFR) [reply include/aligator/core/linesearch-nonmonotone.hpp:27]: Absolutely! I actually do change them in [aligator-bench](https://github.com/Simple-Robotics/aligator-bench).

File: include/aligator/core/linesearch-nonmonotone.hpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+#include <functional>
+
+namespace aligator {
+using proxsuite::nlp::Linesearch;
+using proxsuite::nlp::LinesearchStrategy;
+
+/// @brief Nonmonotone Linesearch algorithm. Modifies the Armijo condition with
+/// a moving average of function values.
+/// @details This is the algorithm from Zhang and Hager, SiOpt 2004.
+template <typename T> struct NonmonotoneLinesearch : Linesearch<T> {
```
[the remark this answers is not in the document]
Comment (ManifoldFR) [reply include/aligator/core/linesearch-nonmonotone.hpp:13]: The linesearch code in proxsuite-nlp is templated on T, I'll make the change to suit this.

File: include/aligator/solvers/proxddp/solver-proxddp.hxx [hunk 4]
```diff
@@ -670,8 +673,10 @@ bool SolverProxDDPTpl<Scalar>::innerLoop(const Problem &problem) {
     Scalar alpha_opt = 1;
     Scalar phi_new;
 
-    switch (sa_strategy) {
-    case StepAcceptanceStrategy::LINESEARCH:
+    switch (sa_strategy_) {
+    case StepAcceptanceStrategy::LINESEARCH_ARMIJO:
```
[the remark this answers is not in the document]
Comment (ManifoldFR) [reply include/aligator/solvers/proxddp/solver-proxddp.hxx:677]: This is intended, both `LINESEARCH_*` should call into the same code here (there's no `break` keyword after `LINESEARCH_ARMIJO`, so it skips to next alternative).

File: tests/python/test_solver.py (hunk 1 above)
Comment (ManifoldFR) [reply tests/python/test_solver.py:63]: Excellent catch; will fix.

**Pull request merged**